Add action to edit any type of file

tgalopin opened the pull request:
In some cases, the usage of `UpdateVersionClass` is not possible as we are dealing with yaml, xml, json, etc. Many different types of configuration exist and this PR aims to provide a generic updater for other types than PHP.

I have two personnal use-cases but I think there are many others:
- working on https://github.com/tgalopin/puli.js, I would like to update my `package.json` on releasing
- working on Symfony applications, I would like to update my assets version (in `app/config/config.yml`) witohut having to use a PHP class

This PR is based on the UpdateVersionClass action, by @dbu :) .

ppetermann commented:
I seem to remember that i changed it, so it can take a file name instead of a classname, like ages ago. Is that not working properly?.

tgalopin commented:
It does, but it's not very clear it does from outside :) . I think having another action is more understandable as UpdateClass on a yml file seems weird IMO. Perhaps should we remove the possibility to provide a file in UpdateClass but then it would break BC.

dbu commented:
> Perhaps should we remove the possibility to
> provide a file in UpdateClass but then it would break BC.

i propose to deprecate using UpdateClass on a non-php file and remove on
the next major version.

is UpdateClass even needed anymore, or could it be completely deprecated
in favor of a general UpdateFileAction?

ppetermann commented:
even if you want to make it more clear (and in my world that would be updating the docs), the code duplication is not necessary, you could just call the existing code.

ppetermann commented:
@dbu breaking bc for pure cosmetics?

dbu commented:
imho its not cosmetics, its very confusing to use an action called UpdateClass to update a yml or json file.

/if/ UpdateClass does nothing special with the fact that its handling a PHP class, i think it should be deprecated and we have just an UpdateFile action that is clear it can update any file. if UpdateClass does something special, we should have two actions (rather than have UpdateClass check if its updating a PHP file or a different file and do different things).

its about developer experience and clarity of the code. and i would not break BC right now, only deprecate whatever is unneeded and when you do a major version bump for some reason, its the time to break BC and tell people how to update.

tgalopin commented:
I fully agree with @dbu, i think having UpdateClass is confusin to edit json/yml files.

However, UpdateClass does have a specific purpose that could be kept: find using Reflection API the class filename. We could have the following comportement:

- UpdateFile accepts any kind of file and reeplace all the occurences of the current version with the next one
- UpdateClass extends UpdateFile and accepts ONLY class names: using Reflection API, it finds the file associated to the class and run UpdateFile on it

What do you think?

dbu commented:
sounds about right to me, yes.

for now, UpdateClass would detect when the "class" is actually a file
and continue to work, but this behaviour would be deprecated and
recommended to use UpdateFile for that use case.

tgalopin commented:
Should I implement that in this PR? How do you deal with depreciations? Using trigger_error?

tgalopin commented:
I updated the code to do the following:

1) when using UpdateClass with a file, a depreciation error will be triggered (using the same conventions as Symfony (see https://github.com/symfony/symfony/pull/13060). In real world cases, it may look like this:

```
Pre-release-actions 
    1) Application version update : New version [1.2.3] updated into /home/tgalopin/devenv/www/RMT/src/Liip/RMT/Application.php: OK
    2) Composer Update Action : OK
    3) Changelog Update Action : OK
    4) Update Version Class Action : PHP Deprecated:  Updating a file using update-version-class action is deprecated and will be removed in 2.0. Use the update-version-file action instead. in /home/tgalopin/devenv/www/RMT/src/Liip/RMT/Action/UpdateVersionClassAction.php on line 55
OK
    5) Vcs Commit Action : OK
```

It sill works and it display a depreciation, which is what we wanted.

2) when using UpdateClass with a class, it works as expected, but I figured that the class needs to be present in the context for the Reflection API to work. Therefore, as RMT is a standalone script, we usually won't have this class available. What do you think of trying to find if the directory of the `.rmt.yml` file contains a `vendor/autoload.php` and load it? It would be more powerful IMO.

3) when using UpdateFile, no specific requirement other than the file exists is needed.

dbu commented on src/Liip/RMT/Action/UpdateVersionFileAction.php:
```diff
@@ -0,0 +1,75 @@
+ *
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+class UpdateVersionFileAction extends BaseAction
```
why call this UpdateVersionFileAction and not just UpdateFileAction?

dbu commented:
i think this makes sense.

if the maintainers of this package agree, we need to also update the documentation to explain how to correctly use those actions now.

tgalopin commented:
I did the changes as following:

- I renamed UpdateVersionFile to UpdateFile
- I created UpdateClass which is the new version of UpdateVersionClass (without the possibility to update anything but PHP classes)
- I updated UpdateVersionClass to trigger a depreciation and to dispatch either to UpdateFIle or UpdateClass according to the type of value we have
- I update ComposerUpdate action as is nothing more than a specific file update (now, it extends UpdateFile action)
- I modified the documentation accordingly

What do you think?

dbu commented:
:+1:

tgalopin commented:
Ping ? :)

dbu commented:
@jeanmonod could you have a look at this please? i am not a maintainer of this package, so i can't decide on this. i think it would make sense like this and be less confusing for the users.

jeanmonod commented:
I will take a look this evening. As a first look, I will say that maybe 1 or 2 functional tests could be great, but that's not an obligation. Do we already have tests for this action?

tgalopin commented:
I saw no test for any action, that's why I didn't do it.

jeanmonod commented on README.md:
```diff
@@ -181,13 +181,20 @@ Actions can be used for pre or post release parts.
+    * Option `pattern`: optional, use to specify the string replacement pattern in your
+      version file. %version% will be replaced by the current / next version strings.
+      For example you could use `const VERSION = '%version%';`. If you do not specify
+      this option, every occurrence of the version string in the file will be replaced.
```
For me this is an issue as you can have many situations where you don't have a current version. For example, when you start an empty project. Or if you are using the `vcs` persister, each time you release a new branch, there is no "current" version.

jeanmonod replied:
Personally I will be more in favor of a regex with a capture group. As an exemple, you provide `/const VERSION = '([^']+)';` and you inject the new version to replace what was inside the group.

jeanmonod commented on src/Liip/RMT/Action/UpdateFileAction.php:
```diff
@@ -0,0 +1,72 @@
+        $content = file_get_contents($filename);
+
+        if (false === strpos($content, $current)) {
+            throw new Exception('The file ' . $filename . ' does not contains the current version ' . $current);
```
This exception will be raised in all case when you have a `%version%` as the current value, isn't it?

jeanmonod commented on src/Liip/RMT/Action/UpdateFileAction.php:
```diff
@@ -0,0 +1,72 @@
+        $content = str_replace($current, $next, $content);
+
+        if (false === strpos($content, $next) || @file_put_contents($filename, $content) === false) {
+            throw new Exception('The file ' . $filename . ' could not be updated with version ' . $next);
```
Here it could be great to split the two tests and to provide two different explicit error messages. It can be very helpful to understand if it's about the content of the file or the access to the file that is wrong

jeanmonod replied:
For the `strpos` check, you have to cast the $next variable. In case of simple version, the provided version is an integer. So you have to do `strpos($content, (string)$next)`

jeanmonod commented:
And about the functional tests. Here is a first basic draft:

```
<?php

namespace Liip\RMT\Tests\Functional;

class UpdateVersionActionTest extends RMTFunctionalTestBase
{
    public function testCommand()
    {
        $this->createChangelog('simple');
        $this->createConfig('simple', 'changelog', array(
            'pre-release-actions' => array(
                'update-file' => array('file' => 'test.yml'),
            ),
        ));
        file_put_contents('test.yml', 'version: 1');

        exec('./RMT release -n --comment=test');
        
        //$this->manualDebug();
        $this->assertEquals('version: 2', file_get_contents('test.yml'));
    }
}
```

You will see that for the moment this test is not passing due to the casting comment just above... To debug it? Just uncomment the commented line ;)

tgalopin commented:
I'll work on that tomorrow :) most of the code was here before my PR though, I just reorganised it :) . But you're right, I'll improve it when I've a bit of time :)

jeanmonod commented:
@tgalopin PING ;) Do you need help?

tgalopin commented:
I don't have much time lately (exams :) ), feel free to work on it :) ! Otherwise, I should be able to work on it in a few weeks

ppetermann commented:
not sure how the custom changelog formatter fits in this PR

tgalopin commented:
Ooops, I didn't plan to push this here. I close this PR but feel free to continue it if you want :) .
